File: service/src/whf/eval/metrics.py
"""Pure metric functions for the evaluation harness. Every function accepts numpy arrays of equal length."""

from __future__ import annotations

import numpy as np


def _arr(*values: object) -> list[np.ndarray]:
    return [np.asarray(v, dtype=float) for v in values]


def mae(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    y, p = _arr(y_true, y_pred)
    return float(np.mean(np.abs(y - p))) if len(y) else float("nan")


def bias(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Mean of forecast minus truth: positive means the forecast runs high."""
    y, p = _arr(y_true, y_pred)
    return float(np.mean(p - y)) if len(y) else float("nan")


def coverage(y_true: np.ndarray, low: np.ndarray, high: np.ndarray) -> float:
    y, lo, hi = _arr(y_true, low, high)
    if len(y) == 0:
        return float("nan")
    return float(np.mean((lo <= y) & (y <= hi)))


def weighted_quantile_loss(y_true: np.ndarray, quantile_preds: dict[float, np.ndarray]) -> float:
    """Mean over quantiles of the scaled pinball loss, as used by the GIFT-Eval benchmark."""
    (y,) = _arr(y_true)
    scale = float(np.mean(np.abs(y))) if len(y) else 0.0
    if scale == 0.0 or not quantile_preds:
        return float("nan")
    losses = []
    for q, pred in quantile_preds.items():
        (p,) = _arr(pred)
        diff = y - p
        losses.append(2.0 * float(np.mean(np.maximum(q * diff, (q - 1.0) * diff))))
    return float(np.mean(losses)) / scale


def overload_precision_recall(true_over: np.ndarray, pred_over: np.ndarray) -> tuple[float, float]:
    t, p = np.asarray(true_over, dtype=bool), np.asarray(pred_over, dtype=bool)
    tp = float(np.sum(t & p))
    precision = tp / float(np.sum(p)) if np.sum(p) else float("nan")
    recall = tp / float(np.sum(t)) if np.sum(t) else float("nan")
    return precision, recall

File: server/forecast-core/src/main/java/com/workloadhub/forecast/eval/Metrics.java
package com.workloadhub.forecast.eval;

/** Pure metric functions over equal-length arrays; NaN on empty input. */
public final class Metrics {

    private Metrics() {
    }

    public static double mae(double[] y, double[] p) {
        return com.workloadhub.forecast.Numbers.mae(y, p);
    }

    /** Mean of forecast minus truth: positive means the forecast runs high. */
    public static double bias(double[] y, double[] p) {
        if (y.length == 0) {
            return Double.NaN;
        }
        double s = 0;
        for (int i = 0; i < y.length; i++) {
            s += p[i] - y[i];
        }
        return s / y.length;
    }

    public static double[] overloadPrecisionRecall(boolean[] trueOver, boolean[] predOver) {
        int tp = 0;
        int predicted = 0;
        int actual = 0;
        for (int i = 0; i < trueOver.length; i++) {
            if (predOver[i]) {
                predicted++;
            }
            if (trueOver[i]) {
                actual++;
            }
            if (trueOver[i] && predOver[i]) {
                tp++;
            }
        }
        return new double[] {predicted == 0 ? Double.NaN : (double) tp / predicted, actual == 0 ? Double.NaN : (double) tp / actual};
    }
}
